optuna/integration/pytorch_lightning.py

```diff
@@ -52,7 +68,39 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
             warnings.warn(message)
             return
 
-        self._trial.report(current_score, step=epoch)
-        if self._trial.should_prune():
+        should_stop = False
+        if trainer.is_global_zero:
+            self._trial.report(current_score, step=epoch)
+            should_stop = self._trial.should_prune()
+        should_stop = trainer.training_type_plugin.broadcast(should_stop)
+        if not should_stop:
+            return
+
+        if not self.is_ddp_backend:
+            message = "Trial was pruned at epoch {}.".format(epoch)
+            raise optuna.TrialPruned(message)
+        else:
+            # stop every ddp process if global rank 0 process decides to stop
+            trainer.should_stop = True
+            if trainer.is_global_zero:
+                self._trial.set_system_attr("pruned", True)
+                self._trial.set_system_attr("epoch", epoch)
+
+    def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        if not self.is_ddp_backend:
+            return
+
+        # Because on_validation_end is executed in spawned processes,
+        # _trial_report is necessary to update the memory in main process, not to update the RDB.
```

Comment:
A nit.

IMO, reporting the values to update the state of trials is a bit hacky.
Not sure, but `trial.report` can be removed if the trial or storage provides a method to synchronize status of trials. But, it is definitely out-of-scope of this PR.
 
```suggestion
        # _trial.report is necessary to update the memory in main process, not to update the RDB.
```